docs/design/features/ScalableApproximateCounting.md

```diff
@@ -0,0 +1,185 @@
+
+Given that we already need to tolerate some inaccuracy, the idea came about to see if we could leverage that to produce an intentionally approximate counting scheme that was nearly as accurate as *interlocked* but with less overhead.
+
+A bit of research turned up an interesting paper by Dice, Lev, and Moir: [Scalable Statistics Counters](https://dl.acm.org/doi/10.1145/2486159.2486182) which was concerned with a similar problem. Here the key insight is that we can leverage randomization to count probabilistically: once the counter value is sufficiently large, instead of always trying to update its value by 1 for each increment, we update it by $N$ with probability $1/N$. So as the counter value grows we update it less and less often, thus reducing the amount of contention counter updates cause, and the counter's expected value is the correct count, with some controllable likelihood of the count being too large or to small. Because the total number of updates is limited, we can use *interlocked* operations for the updates without unduly harming scalability.
+
+The approach in the paper relies either on using standard floating-point representations for counts or a "software float" representation where part of the counter storage is an exponent and the remainder a mantissa. Those representations were not well suited for our work, so we set about to find a similar scheme that would work for regular integer data types.
+
+The key benefit of the exponent/mantissa form is that it makes it relatively simple to estimate the magnitude of the counter&mdash;the exponent is the integer part of the $\log_2$ of the counter value, and in particular, to compute an approximately correct quantity $1/N$ quickly without needing to divide if we restrict $N$ to a power of 2. For normal integer data we can compute the same thing by simply finding the highest set bit of the data, and can check the probability by computing the remainder left by dividing some random value by simple masks and compares.
+
+## The Scalable Counter
+
+Here is a C# implementation of our counter:
+```C#
+    static void ScalableProfileCount(ref uint counter)
+    {
+        uint count = counter;
+        uint delta = 1;
+
+        if (count > 0)
+        {
+            int logCount = 31 - (int) uint.LeadingZeroCount(count);
+
+            if (logCount >= 13)
```

Comment:
can we have a sort of fast path here? E.g. 
```
if (count & 3FFFF) // is counter already large enough?
{
    int logCount = 31 - (int) uint.LeadingZeroCount(count);
    ..
}
```
e.g. LZC is not cheap on arm since there is no scalar version

Reply:
Moreover, we can probably inline it in JIT codegen if needed

Reply:
You confused me for a second commenting on the version in the doc and not the code in the runtime itself. 

Yes we can do some sort of check like this.